apps/omg_watcher/lib/exit_processor/in_flight_exit_info.ex

```diff
@@ -0,0 +1,222 @@
+          timestamp: timestamp,
+          contract_id: contract_id,
+          is_active: is_active
+        }
+      }
+    end
+  end
+
+  def make_db_update({_ife_hash, %__MODULE__{} = _ife} = update) do
+    {:put, :in_flight_exit_info, update}
+  end
+
+  @spec piggyback(t(), non_neg_integer()) :: {:ok, t()} | {:error, :non_existent_exit | :cannot_piggyback}
+  def piggyback(ife, index)
+
+  def piggyback(%__MODULE__{exit_map: exit_map} = ife, index) when index in @output_index_range do
+    with exit <- Map.get(exit_map, index),
+         {:ok, updated_exit} <- piggyback_exit(exit) do
+      {:ok, %{ife | exit_map: Map.merge(exit_map, %{index => updated_exit})}}
+    end
+  end
+
+  def piggyback(%__MODULE__{}, _), do: {:error, :non_existent_exit}
+
+  defp piggyback_exit(%{is_piggybacked: false, is_finalized: false}),
+    do: {:ok, %{is_piggybacked: true, is_finalized: false}}
+
+  defp piggyback_exit(_), do: {:error, :cannot_piggyback}
+
+  @spec challenge(t(), non_neg_integer()) :: {:ok, t()} | {:error, :cannot_challenge}
+  def challenge(ife, competitor_position)
+
+  def challenge(%__MODULE__{oldest_competitor: nil} = ife, competitor_position),
+    do: %{ife | is_canonical: false, oldest_competitor: decode_tx_position(competitor_position)}
+
+  def challenge(%__MODULE__{oldest_competitor: current_oldest} = ife, competitor_position) do
+    with decoded_competitor_pos <- decode_tx_position(competitor_position),
+         true <- is_older?(decoded_competitor_pos, current_oldest) do
+      %{ife | is_canonical: false, oldest_competitor: decoded_competitor_pos}
+    else
+      _ -> {:error, :cannot_challenge}
+    end
+  end
+
+  @spec challenge_piggyback(t(), integer()) :: {:ok, t()} | {:error, :non_existent_exit | :cannot_challenge}
+  def challenge_piggyback(ife, index)
+
+  def challenge_piggyback(%__MODULE__{exit_map: exit_map} = ife, index) when index in @output_index_range do
+    with %{is_piggybacked: true, is_finalized: false} <- Map.get(exit_map, index) do
+      {:ok, %{ife | exit_map: Map.merge(exit_map, %{index => %{is_piggybacked: false, is_finalized: false}})}}
+    else
+      _ -> {:error, :cannot_challenge}
+    end
+  end
+
+  def challenge_piggyback(%__MODULE__{}, _), do: {:error, :nxon_existent_exit}
+
+  @spec respond_to_challenge(t(), pos_integer()) :: {:ok, t()} | :error
+  def respond_to_challenge(ife, tx_position)
+
+  def respond_to_challenge(%__MODULE__{oldest_competitor: nil, tx_pos: nil} = ife, tx_position) do
+    decoded = decode_tx_position(tx_position)
+    {:ok, %{ife | oldest_competitor: decoded, is_canonical: true, tx_pos: decoded}}
+  end
+
+  def respond_to_challenge(%__MODULE__{oldest_competitor: current_oldest, tx_pos: nil} = ife, tx_position) do
+    decoded = decode_tx_position(tx_position)
+
+    if is_older?(decoded, current_oldest) do
+      {:ok, %{ife | oldest_competitor: decoded, is_canonical: true, tx_pos: decoded}}
+    else
+      :error
```

Comment:
It is better to use tuple errors. If user does not care what type of error is it, she just matches first element of tuple.